File: examples/qualifiers/kotlin/core/console.kt
package salvo.core.console

import salvo.core.string.*

interface Console {
    fun print(message: String)
}

class __Mon_Console(private val inner: Console) : Console {
    override fun print(message: String) =
        synchronized(inner) { inner.print(message) }
}

class StdOutConsole : Console {
    override fun print(message: String) {
        kotlin.io.print(message)
    }
}

fun println(console: Console, message: String) {
    console.print(message)
    console.print("\n")
}

File: examples/collections/rust/core/console.rs
use crate::core_iterator::*;
use crate::core_string::*;

pub trait Console {
    fn print(&mut self, message: &String);
}

#[derive(Clone)]
pub struct __Mon_Console {
    inner: std::sync::Arc<std::sync::Mutex<dyn Console + Send>>,
}

impl __Mon_Console {
    pub fn new(inner: std::sync::Arc<std::sync::Mutex<dyn Console + Send>>) -> Self {
        Self { inner }
    }
}

impl Console for __Mon_Console {
    fn print(&mut self, message: &String) {
        self.inner.lock().unwrap().print(message)
    }
}

pub struct StdOutConsole {
}

impl StdOutConsole {
    pub fn new() -> Self {
        Self { }
    }
}

impl Console for StdOutConsole {
    fn print(&mut self, message: &String) {
        print!("{}", message)
    }
}

pub fn println(console: &mut dyn Console, message: &String) {
    console.print(message);
    console.print(&("\n".to_string()));
}
